scipy/integrate/_tanhsinh.py

```diff
@@ -0,0 +1,755 @@
+    axis_kwargs = dict(axis=-1, keepdims=True)
+
+    # With a jump start (starting at level higher than 0), we haven't
+    # explicitly calculated the integral estimate at lower levels. But we have
+    # all the function value-weight products, so we can compute the
+    # lower-level estimates.
+    if work.Sk.shape[-1] == 0:
+        h = 2 * work.h  # step size at this level
+        n_x = indices[work.n]  # number of abscissa up to this level
+        # The right and left fjwj terms from all levels are concatenated along
+        # the last axis. Get out only the terms up to this level.
+        fjwj_rl = work.fjwj.reshape(n_active, 2, -1)
+        fjwj = fjwj_rl[:, :, :n_x].reshape(n_active, 2*n_x)
+        # Compute the Euler-Maclaurin sum at this level
+        Snm1 = (special.logsumexp(fjwj, **axis_kwargs) + np.log(h) if work.log
+                else np.sum(fjwj, **axis_kwargs) * h)
+        work.Sk = np.concatenate((Snm1, work.Sk), axis=-1)
+
+    if work.n == 1:
+        nan = np.full_like(work.Sn, np.nan)
+        return nan, nan
+
+    # The paper says not to calculate the error for n<=2, but it's not clear
+    # about whether it starts at level 0 or level 1. We start at level 0, so
+    # why not compute the error beginning in level 2?
+    if work.Sk.shape[-1] < 2:
+        h = 4 * work.h  # step size at this level
+        n_x = indices[work.n-1]  # number of abscissa up to this level
+        # The right and left fjwj terms from all levels are concatenated along
+        # the last axis. Get out only the terms up to this level.
+        fjwj_rl = work.fjwj.reshape(len(work.Sn), 2, -1)
+        fjwj = fjwj_rl[..., :n_x].reshape(n_active, 2*n_x)
+        # Compute the Euler-Maclaurin sum at this level
+        Snm2 = (special.logsumexp(fjwj, **axis_kwargs) + np.log(h) if work.log
+                else np.sum(fjwj, **axis_kwargs) * h)
+        work.Sk = np.concatenate((Snm2, work.Sk), axis=-1)
+
+    Snm2 = work.Sk[..., -2]
+    Snm1 = work.Sk[..., -1]
+
+    e1 = np.finfo(work.dtype).eps
+
+    if work.log:
+        log_e1 = np.log(e1)
+        # Currently, only real integrals are supported in log-scale. All
+        # complex values have imaginary part in increments of pi*j, which just
+        # carries sign information of the original integral, so use of
+        # `np.real` here is equivalent to absolute value in real scale.
+        d1 = np.real(special.logsumexp([work.Sn, Snm1 + work.pi*1j], axis=0))
+        d2 = np.real(special.logsumexp([work.Sn, Snm2 + work.pi*1j], axis=0))
+        d3 = log_e1 + np.max(np.real(work.fjwj), axis=-1)
+        d4 = work.d4
+        aerr = np.max([d1 ** 2 / d2, 2 * d1, d3, d4], axis=0)
+        rerr = np.maximum(log_e1, aerr - np.real(work.Sn))
+    else:
+        # Note: explicit computation of log10 of each of these is unnecessary.
+        d1 = np.abs(work.Sn - Snm1)
+        d2 = np.abs(work.Sn - Snm2)
+        d3 = e1 * np.max(np.abs(work.fjwj), axis=-1)
+        d4 = work.d4
```

Comment:
> I don't think `d3` and `d4` should be weighted by $h$. 

From a practical or theoretical standpoint? 

Since these terms effect on the integral estimate are weighted by $h$, I don't see how their effect on the error could be independent of $h$.

But if you mean practically speaking they are so crude that we shouldn't take that into consideration, I understand.

Reply:
OK, so for why we don't include $h$, $d_3$ (and also $d_4$) is just supposed to be a crude lower bound due to arithmetic error due to calculating to only a finite precision. I think the confusion is because in arbitrary precision arithmetic, $\epsilon_1$ has a different meaning from the `eps` you're using with finite-precision arithmetic. I think $\epsilon_1$ is the level of precision that one seeks for all arithmetic operations, and one tries to stay within this precision by performing calculations at a still higher level of precision. In your case `eps` is the tolerance you're seeking for the integral, but it's not a bound for all arithmetic operations. I think to make the lower bounds work in these cases, you should actually take $\epsilon_1$ equal to half of machine $\epsilon$.

Reply:
> In your case `eps` is the tolerance you're seeking for the integral

What is `eps`? 

`e1 = np.finfo(work.dtype).eps`

is machine precision $\epsilon$. By 

> you should actually take $\epsilon_1$ equal to half of machine $\epsilon$.

do you mean that we should take $\epsilon_1 = \sqrt{\epsilon}$?

Most arithmetic options are not accuarate to $\epsilon_1$, but they are typically much better than $\sqrt{\epsilon}$.

---

The number of terms in the EM sum is $n = c / h$, where $c$ is a small constant that we know. Then if $\epsilon_1$ is the precision (worst case relative error of any term), isn't 

$\epsilon_1 c \max_j{\left|(f(x_j)\right|w_j}$ 

an (unnecessarily) conservative *upper* bound of the arithmetic error in calculating the EM sum? It assumes that the error of every term is as large as the error of the largest term, and it assumes that all the errors add together rather than partially cancelling each other out. We could easily (more easily, really) compute a better upper bound on this sort of error as 

$\epsilon_1 \sum_{j\in 1, ..., n} {h \left|(f(x_j)\right|w_j}$

right?